### ISSUE TITLE
GUI: FED:  Server Control - Firmware Update - Full Redfish Support

### ISSUE THREAD

**Issue by derick-montague** [ibm-openbmc/dev#1030]
## Stakeholders
BMC SME: @anoo1
PHYP SME: BJ Vette
RAS SME: Justin Thaler
DR: @nicoleconser
FED: TBD

## Overview
Create the firmware update page using Vue.js and the Redfish API. Users will be able to see the firmware version for running and backup. Users will be able to upload and activate images from their local workstation or TFTP server. Users will be able to reboot BMC from this page.

## UX flow/ Interaction requirements

- [ ]  Users can upload a .tar file from Workstation or enter TFTP server information to upload file
- [ ]  Users can upload then activate the image **check final backend flow**
- [ ]  Users can copy running firmware version to backup when clicking Copy to backup

## InVision link
https://ibm.invisionapp.com/share/4XNZ0JAMJ7B#/screens/319207131_4-8-S-2A_TFTP_Server

## cURL
```
N/A
```
https://github.com/openbmc/docs/blob/master/REDFISH-cheatsheet.md#firmware-update
## Notes

- Do not need to develop Next boot image – #2207 
- This will be for single file upload only. The design for two file image will not be used or needed for this story. [**Help wanted**](https://github.com/openbmc/webui-vue/projects/2#card-38551506)

## Screenshots
Vue
<img width="1074" alt="Screen Shot 2020-05-15 at 10 45 12 AM" src="https://user-images.githubusercontent.com/8097306/82080363-43cea400-9699-11ea-9e99-6dd9dfd8aa41.png">

Angular (for reference)
![Screen Shot 2020-05-15 at 10 39 55 AM](https://user-images.githubusercontent.com/8097306/82079926-83e15700-9698-11ea-9269-3e6e7cfbd922.png)

## References
[Redfish workflow change conversation in slack](https://ibm-systems-power.slack.com/archives/G3WAVQYTH/p1584569722242800)

**Comment by derick-montague** [comment]
This will be done as part of the single firmware upload. May be able to close this story.

**Comment by yoshiemuranaka** [comment]
Latest InVision prototype 
https://ibm.invisionapp.com/d/main#/console/15364725/319627530/preview

**Comment by yoshiemuranaka** [comment]
![Screen Shot 2020-07-30 at 1 52 20 PM](https://user-images.githubusercontent.com/8097306/88973423-e7c4f600-d26b-11ea-846f-71f9ba80ebd0.png)

Gunnar suggest rewording to:

Switch and reboot BMC from backup image <- Title
Image to switch to <- smaller text
OP940 <- pull from backup image
Switch and Reboot BMC  <- Button (edited) 

Include "Switch" or "Change"

**Comment by derick-montague** [comment]
I can see that we could make some changes for clarity. I think these are a step in the right direction. I have some concerns about translation. Since the only change we can make is to boot from the previous image and this page is not just a simple Reboot BMC, it is very important that we make that clear. 

However, I don't think this is just Rebooting the BMC is it? This will also require shutting down the host, correct?

**Comment by yoshiemuranaka** [comment]
>I am concerned about the word switch since we are not switching, we are replacing. The current image is not going to become the backup, correct?

The current image does become backup.
The current and backup are being switched when you reboot from backup image. It is not being replaced.

**Comment by derick-montague** [comment]
> > I am concerned about the word switch since we are not switching, we are replacing. The current image is not going to become the backup, correct?
> 
> The current image does become backup.
> The current and backup are being switched when you reboot from backup image. It is not being replaced.

Really? That doesn't seem correct. If i have a running image and it is not functioning and I determine to run from the back up, the back up should not be replaced with an image that has issues. If that is the case, then we are going to have to acknowledge that to maintain the systems state the user will have to:

1. Reboot the BMC from backup
2. After complete, use the `Copy to backup` to reset the system state

## Questions
1. Is it intuitive to the user what has happened?
    - If not, How do we make that user aware of that?

**Comment by derick-montague** [comment]
Would something like this be more intuitive? 
![image](https://user-images.githubusercontent.com/1183515/89203135-a8ebb480-d579-11ea-9179-7cb7150db9c0.png)

**Comment by derick-montague** [comment]
This potentially could be `swap` or `switch` instead of update. We just need it to be clear what is happening.

**Comment by yoshiemuranaka** [comment]
<img width="319" alt="Screen Shot 2020-08-04 at 8 06 21 AM" src="https://user-images.githubusercontent.com/8097306/89310441-8e711400-d629-11ea-85ef-3299af081ce8.png">

My suggested copy updates ^

**Comment by derick-montague** [comment]
We should make these changes and push it to the design in progress in the webui-vue repo and test with users. I have concerns and I believe we should retest with users.

**Comment by derick-montague** [comment]
I talked with Stacia from the ID team and `switch` is an concern for translation. Here are the edits that I am suggesting.

- Replace the word `switch` with `change`
- `Image to change to`
- `Replace backup image with current image. Current image will become backup image`

**Comment by derick-montague** [comment]
Is there a reason we could add an alert here rather than hiding something that could be critical information to the user in an info icon?

**Comment by ParishrutB** [comment]
How about this:

<img width="307" alt="image" src="https://user-images.githubusercontent.com/61419136/89317619-3d9a0380-d69b-11ea-8e66-582d14ef7ef8.png">

Title reflects the user intention. Help text explains the exact process. Action button maps to both. Removed the 'Next boot image' part as that was just creating more confusion than helping.

**Comment by ParishrutB** [comment]
@derick-montague +1 that a broken current image should not be copied to backup. But won't the user realize that on their own?

**Comment by yoshiemuranaka** [comment]
Also want to point out additional information in confirmation modal

https://ibm.invisionapp.com/d/main#/console/15364725/319627545/preview

<img width="644" alt="Screen Shot 2020-08-04 at 10 55 58 AM" src="https://user-images.githubusercontent.com/8097306/89327707-28908680-d641-11ea-9b59-4d24506774e9.png">

**Comment by derick-montague** [comment]
> @derick-montague +1 that a broken current image should not be copied to backup. But won't the user realize that on their own?

They have no option. When changing from backup to current, the images swap. They will have to `Copy current to backup`

**Comment by derick-montague** [comment]
@yoshiemuranaka is it true that the user will have to login again. That is not true when they reboot the BMC.

![image](https://user-images.githubusercontent.com/1183515/89349871-8c807280-d674-11ea-9347-b2361f022df5.png)

**Comment by derick-montague** [comment]
![image](https://user-images.githubusercontent.com/1183515/89350028-dbc6a300-d674-11ea-9014-8b8c689d29d9.png)

**Comment by nicoleconser** [comment]
@ParishrutB 2 suggestions: specify the object in the button text, and spell out "and" — so **Reboot BMC and change images**

@derick-montague Did Stacia say anything about "swap"? 

**Reboot BMC and swap images** seems more accurate (and less ambiguous) than "change."

**Comment by derick-montague** [comment]
> @derick-montague Did Stacia say anything about "swap"?

No, we only talked about switch

> Reboot BMC and swap images seems more accurate (and less ambiguous) than "change."

+1 - This is an accurate description. Using it as a verb here makes sense.

>  In a system with virtual storage, to write the main storage image of a job to auxiliary storage and read the image of another job into processor storage.
>
> Source: Dictionary of Printing, Scale Out Network Attached Storage, Terminology Extraction Process, Translation Services Center
> Subject area: Information technology // Software // Cloud and IT optimization // Storage management, backup and recovery

**Comment by derick-montague** [comment]
Although something like this may work, there are some other possible options we should think about. Might be a good topic for a short discussion. 

![image](https://user-images.githubusercontent.com/1183515/89353386-e2a4e400-d67b-11ea-9c38-32fc80a57f5d.png)

**Comment by yoshiemuranaka** [comment]
>is it true that the user will have to login again. That is not true when they reboot the BMC.

Not that I know of from testing so far. Maybe we can use similar language to what we have on the Reboot BMC Page.

>When you reboot the BMC, your web browser loses contact with the BMC for several minutes. When the BMC is back online, you may need to log in again.

I have implemented something similar in the code, was going to bring up during our next playback

**Comment by ParishrutB** [comment]
> Although something like this may work, there are some other possible options we should think about. Might be a good topic for a short discussion.
> 
> ![image](https://user-images.githubusercontent.com/1183515/89353454-08ca8400-d67c-11ea-9869-32061463c0b3.png)

Does "Change to backup image" in title and "Swap images" in the button align/match with each other? If we are deciding on 'swap' then we should use it in title too. 

Accuracy is subjective and we may need to check with the users. This question is ideally answered by them.

**Comment by ParishrutB** [comment]
<img width="247" alt="image" src="https://user-images.githubusercontent.com/61419136/89430023-ead45080-d75b-11ea-978c-ba20c456b966.png">

Checking if this much works. If yes, then we just need to discuss about the title.

**Comment by ParishrutB** [comment]
> @yoshiemuranaka is it true that the user will have to login again. That is not true when they reboot the BMC.
> 
> ![image](https://user-images.githubusercontent.com/1183515/89349871-8c807280-d674-11ea-9347-b2361f022df5.png)

This is true. I saw it in the test on witherspoon.

**Comment by derick-montague** [comment]
I don't believe it is true. We do not have to login when we reboot the BMC, so I don't think this is any different. I believe that that verbiage in the old UI is not accurate. @gtmills, right? We had this conversation with Ed Tenuous last year about how our GUI does not have that requirement, which is a differentiator from competitor GUIs.

**Comment by yoshiemuranaka** [comment]
MVP 

Title: Change to backup image
Label: Backup image
Button: Change image and reboot BMC

<img width="367" alt="Screen Shot 2020-08-06 at 10 15 36 AM" src="https://user-images.githubusercontent.com/8097306/89561742-cdd86580-d7cd-11ea-8aca-eb23e73e4491.png">

**Comment by yoshiemuranaka** [comment]
Potential direction to explore in next iteration @ParishrutB 

<img width="1792" alt="Screen Shot 2020-08-06 at 10 10 51 AM" src="https://user-images.githubusercontent.com/8097306/89561975-2576d100-d7ce-11ea-90a9-269099efa480.png">

**Comment by yoshiemuranaka** [comment]
Need final copy for three toast notifications

1. First (bottom)- when user clicks `Upload and reboot BMC`
2. Second (middle) - when upload step complete 
3. Third (top) - hardcoded (6 minute) timeout

@derick-montague @ParishrutB 


<img width="782" alt="Screen Shot 2020-08-06 at 12 07 43 PM" src="https://user-images.githubusercontent.com/8097306/89572223-adb0a280-d7dd-11ea-9580-683580c4f271.png">

**Comment by derick-montague** [comment]
![image](https://user-images.githubusercontent.com/1183515/89588574-88398e00-d809-11ea-9fd6-c6da75aa7df4.png)
### Code update started
The upload was successful. During code update, the BMC will be not be responsive. Wait for the code update notification before making any changes.

---

![image](https://user-images.githubusercontent.com/1183515/89589111-9a67fc00-d80a-11ea-9eda-fb0f206b78c5.png)

###  Verify code update
Refresh the application to confirm the code update has completed and was successful.

**Comment by yoshiemuranaka** [comment]
Resolved by https://gerrit.openbmc-project.xyz/c/openbmc/webui-vue/+/34744

**Comment by derick-montague** [comment]
Images for testing: https://ibm-systems-power.slack.com/archives/GQR6QQXT8/p1597076066001500

**Comment by rfrandse** [comment]
https://gerrit.openbmc-project.xyz/c/openbmc/webui-vue/+/34744 merged
